widget_weather.go

```diff
@@ -0,0 +1,226 @@
+package main
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"image"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/muesli/streamdeck"
+)
+
+//go:embed assets/weather
+var weatherImages embed.FS
+
+func weatherImage(name string) image.Image {
+	b, err := weatherImages.ReadFile(name)
+	if err != nil {
+		panic(err)
+	}
+
+	icon, _, err := image.Decode(bytes.NewReader(b))
+	if err != nil {
+		panic(err)
+	}
+
+	return icon
+}
+
+// WeatherWidget is a widget displaying the current weather.
+type WeatherWidget struct {
+	*ButtonWidget
+
+	data  WeatherData
+	theme string
+}
+
+// WeatherData handles fetches and parsing weather data.
+type WeatherData struct {
+	location string
+	unit     string
+
+	refresh time.Time
+
+	response      string
+	responseMutex sync.RWMutex
+}
+
+// Condition returns the current condition.
+func (w *WeatherData) Condition() (string, error) {
+	w.responseMutex.RLock()
+	defer w.responseMutex.RUnlock()
+
+	if strings.Contains(w.response, "Unknown location") {
+		fmt.Println("unknown location:", w.location)
+		return "", nil
+	}
+
+	wttr := strings.Split(w.response, " ")
+	if len(wttr) != 2 {
+		return "", fmt.Errorf("can't parse weather response: %s", w.response)
+	}
+
+	return wttr[0], nil
+}
+
+// Temperature returns the current temperature.
+func (w *WeatherData) Temperature() (string, error) {
+	w.responseMutex.RLock()
+	defer w.responseMutex.RUnlock()
+
+	if strings.Contains(w.response, "Unknown location") {
+		fmt.Println("unknown location:", w.location)
+		return "", nil
+	}
+
+	wttr := strings.Split(w.response, " ")
+	if len(wttr) != 2 {
+		return "", fmt.Errorf("can't parse weather response: %s", w.response)
+	}
+
+	return strings.Replace(wttr[1], "+", "", 1), nil
+}
+
+// Ready returns true when weather data is available.
+func (w *WeatherData) Ready() bool {
+	w.responseMutex.RLock()
+	defer w.responseMutex.RUnlock()
+
+	return len(w.response) > 0
+}
+
+// Fetch retrieves weather data when required.
+func (w *WeatherData) Fetch() {
+	w.responseMutex.Lock()
+	defer w.responseMutex.Unlock()
+
+	if time.Since(w.refresh) < time.Minute*15 {
+		return
+	}
+	// fmt.Println("Refreshing weather data...")
+
+	url := "http://wttr.in/" + w.location + "?format=%x+%t" + formatUnit(w.unit)
+
+	resp, err := http.Get(url) //nolint:gosec
+	if err != nil {
+		fmt.Println("can't fetch weather data:", err)
+		return
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Println("can't read weather data:", err)
+		return
+	}
+
+	w.refresh = time.Now()
+	w.response = string(body)
+}
+
+// NewWeatherWidget returns a new WeatherWidget.
+func NewWeatherWidget(bw *BaseWidget, opts WidgetConfig) (*WeatherWidget, error) {
+	bw.setInterval(opts.Interval, 60000)
```

Comment:
As far as I tested this does not change the update interval... I'm still debugging but there seems to be something wrong

Reply:
I see what's happening: `NewButtonWidget` overrides this by calling `bw.SetInterval` itself. Fixing.

Reply:
Ok, pushed a fix. This also fixes not drawing the background initially while waiting for the first weather data response. I'm not too fond of how I fixed it, however. Will think of a nicer solution.